---
title: "Accuracy of gene expression measurements"
author: "Jackie Roger & Holly Beale"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document: default
  # pdf_document: default
  # fig_caption: yes
---

```{r setup, include = FALSE}

library(readr)
library(jsonlite)
library(magrittr)
library(dplyr)
library(tidyr)
library(knitr)

```

### Background

The purpose of this Rmd is to quantify the accuracy of outlier leads.

### User input: sample id, outlier lead & sequencing depth
##### (measured in millions of mapped, exonic, non-duplicate (MEND) reads)

```{r echo = TRUE}

sample_id <- "TH34_1447_S01"
outlier_lead <- "PARP1"
mend_depth <- 51.5

```

### Read in data

```{r echo = TRUE}

# Sample-specific data

outlier_info <- read_tsv(paste0("data/sample_in/", sample_id, "/outlier_results_", sample_id), col_types = cols())
threshold_info <- fromJSON(paste0("data/sample_in/", sample_id, "/3.json"))

# Non-sample-specific data

per_gene_expression_info <- read_tsv(paste0("data/in/per_gene_expression.tsv.gz"), col_types = cols())
cohort_names_info <- read_tsv(paste0("data/in/cohort_names.tsv"), col_types = cols())

```

### Figure out thresholds, expression bin & mend depth bin

```{r echo = TRUE}

# Figure out outlier thresholds

cohort_names <- cohort_names_info$useful_name

thresholds <- lapply(cohort_names, function(cohort) {
  these_thresholds <- threshold_info[[cohort]]
  this_threshold <- these_thresholds$high[outlier_lead]
  if (is.null(this_threshold)) this_threshold <- NA
  return (this_threshold)
  }) %>%
  unlist %>%
  as.numeric

# Figure out expression bin

exp <- outlier_info %>% filter(Gene == outlier_lead) %>% pull(sample)

expression_bins_all <- c("0-1", "1-3", "3-5", "5-7", "7-10", "10-15")
exp_bins <- strsplit(expression_bins_all, '-')

expression_bin <- expression_bins_all[lapply(exp_bins, function(exp_bin) {
  exp > as.numeric(exp_bin[1]) & exp < as.numeric(exp_bin[2])
  }) %>%
  unlist]

# Figure out MEND depth bin

mend_depth_bins_all <- seq(0, 44, 4)

if (mend_depth >= 42) mend_depth_bin <- 44 else mend_depth <- round(mend_depth / 4) * 4

```

### Get all relevant accuracy measurements for each threshold

```{r echo = TRUE}

# Start building summary table

summary <- tibble(sample = sample_id,
                  gene = outlier_lead,
                  cohort = cohort_names,
                  expression = exp,
                  threshold = thresholds)

# Figure out if expression exceeds each threshold & calculate percent difference

summary <- summary %>%
  mutate(exceeds_threshold = (expression > threshold),
         percent_difference = 100 * ((abs(expression - threshold)) / expression))

# Get accuracy distributions

measurement_stats_all <- lapply(summary$percent_difference, function(threshold) {
  if (!(is.na(threshold))) {
    # Set accuracy threshold and get accuracy info
    non_zero_abundance <- per_gene_expression_info %>%
      dplyr::filter(f_expression_at_max_depth_Gt0)  %>% 
      mutate(within_accuracy_threshold_of_deepest_val =
               expression > (1 - threshold / 100) * expression_at_max_depth &
               expression < (1 + threshold / 100) * expression_at_max_depth)
    # Make expression bins and generate summary statistics
    break_sequence <- c(0, 1, 3, 5, 7, 10, 20, 30, 50,
                        ceiling(max(non_zero_abundance$expression_at_max_depth)))
    set.seed(1)
    abundance_by_expression_bin <- non_zero_abundance %>% 
      ungroup() %>% 
      mutate(expression_bin = cut(expression_at_max_depth, breaks = break_sequence, dig.lab = 6))
    abundance_stats <- abundance_by_expression_bin	%>% 
      group_by(expression_bin, bin_median, UMEND, parent_id ) %>%
      summarize(pct_accurately_measured = sum(within_accuracy_threshold_of_deepest_val) / n(),
                n_genes_in_bin = length(unique(gene)),
                expression_bin_label = paste0(gsub(
                  "]", "", gsub(",", "-", gsub("[\\(]*", "", unique(expression_bin)))), " (n=", n_genes_in_bin, ")")) 
    abundance_stats <- abundance_stats %>%
      mutate(min_bin_label = gsub(" .*$", "", expression_bin_label),
             pct_accurately_measured = pct_accurately_measured * 100)
    return(abundance_stats)
  }
})

# Filter results

measurement_stats_filtered <- lapply(measurement_stats_all, function(msa) {
  if (!(is.null(msa))) {
    exp <- expression_bin # for some reason it doesn't work without doing it this way
    stats_filtered <- msa %>%
      filter((bin_median == mend_depth_bin) && (min_bin_label == exp)) %>%
      arrange(parent_id)
    return(stats_filtered)
  }
})

```

### Calculate gene expression measurement accuracy stats

```{r echo = TRUE, warning = FALSE}

# Get accuracy stats

get_stats <- function(fun, s) {
  stats = lapply(measurement_stats_filtered, function(msf) {
    if (!(is.null(msf))) {
      if (!(is.na(fun))) {
        stat <- fun(msf$pct_accurately_measured)
      } else if (!(is.na(s))) {
        stat <- msf$pct_accurately_measured[s]
      }
      return (stat)
    } else {
      return (NA)
    }
  }) %>%
    unlist %>%
    as.numeric
  return (stats)
}

# Add accuracy stats to summary table

summary <- summary %>%
  mutate(
    avg = get_stats(mean, NA),
    min = get_stats(min, NA),
    max = get_stats(max, NA),
    S1 = get_stats(NA, 1),
    S2 = get_stats(NA, 2),
    S3 = get_stats(NA, 3),
    S4 = get_stats(NA, 4),
    S5 = get_stats(NA, 5)
  )

```

### Write summary to tsv and show tables

```{r echo = TRUE}

# Clean up names for writing to outfile

summary <- summary %>%
  rename("Sample" = sample,
         "Gene" = gene,
         "Cohort" = cohort,
         "Expression" = expression,
         "Threshold" = threshold,
         "Exceeds Threshold" = exceeds_threshold,
         "Percent Difference" = percent_difference,
         "Mean Accuracy" = avg,
         "Minimum Accuracy" = min,
         "Maximum Accuracy" = max,
         "Sample 1 Accuracy" = S1,
         "Sample 2 Accuracy" = S2,
         "Sample 3 Accuracy" = S3,
         "Sample 4 Accuracy" = S4,
         "Sample 5 Accuracy" = S5)

summary$Cohort <- cohort_names_info$pretty_name[match(summary$Cohort, cohort_names_info$useful_name)]

# Write to outfile

write_tsv(summary, paste0("results/", sample_id, "/sample_", sample_id, "__gene_", outlier_lead, "__summary.tsv"))

# Show short summary table & long summary table

kable(summary[1:8], digits = 2)
kable(summary, digits = 2)

```